svm/tests/integration_test.rs

```diff
@@ -251,8 +257,42 @@ impl SvmTestEnvironment<'_> {
         let mut mock_bank_accounts = self.mock_bank.account_shared_data.write().unwrap();
         mock_bank_accounts.extend(final_accounts_actual);
 
+        // update global program cache
+        for processing_result in batch_output.processing_results.iter() {
+            if let Some(ProcessedTransaction::Executed(executed_tx)) =
+                processing_result.processed_transaction()
+            {
+                let programs_modified_by_tx = &executed_tx.programs_modified_by_tx;
+                if executed_tx.was_successful() && !programs_modified_by_tx.is_empty() {
+                    self.batch_processor
+                        .program_cache
+                        .write()
+                        .unwrap()
+                        .merge(programs_modified_by_tx);
+                }
+            }
+        }
+
         batch_output
     }
+
+    pub fn is_program_blocked(&self, program_id: &Pubkey) -> bool {
+        let program_cache_entries = self
+            .batch_processor
+            .program_cache
+            .read()
+            .unwrap()
+            .get_flattened_entries_for_tests();
+
+        // in the same entry, a new valid loaderv3 program may have a Loaded entry with a later execution slot
+        // in a later entry, the same loaderv3 program will have a DelayedVisibility tombstone
+        // a new loaderv1/v2 account will have a FailedVerification tombstone
+        // and a closed loaderv3 program or any loaderv3 buffer will have a Closed tombstone
+        // we must search because entries are unsorted and old entries persist
+        program_cache_entries.iter().any(|(key, entry)| {
+            key == program_id && (entry.effective_slot > EXECUTION_SLOT || entry.is_tombstone())
+        })
+    }
```

Comment:
i was not smart enough to figure out the index structs/enums that lets you get only the latest entry

Reply:
nevermind i can just do this right ill put it back in review in a minute